src/main/scala/org/ergoplatform/settings/Parameters.scala

```diff
@@ -1,30 +1,283 @@
   */
-object Parameters {
-
-  // Max total computation cost of a block.
-  val MaxBlockCost: Long = 1000000
+class Parameters(val height: Height, val parametersTable: Map[Byte, Int]) {
 
-  // Max size of transactions section of a block.
-  val MaxBlockSize: Int = 512 * 1024
+  import Parameters._
 
-  val Kdefault = 1250000
-  val Kmax = 2500000
-  val Kmin = 0
-  val Kstep = 50000
-
-  /** Cost of storing 1 byte per Constants.StoragePeriod blocks, in nanoErgs
+  /**
+    * Cost of storing 1 byte per Constants.StoragePeriod blocks, in nanoErgs.
     */
-  val K: Long = Kdefault
+  lazy val storageFeeFactor: Int = parametersTable(StorageFeeFactorIncrease)
 
   /** To prevent creation of dust which is not profitable to charge storage fee from, we have this min-value per-byte
     * parameter.
     */
-  val MinValuePerByte: Long = 30 * 12
+  lazy val minValuePerByte: Int = parametersTable(MinValuePerByteIncrease)
+
+  /**
+    * Max size of transactions section of a block.
+    */
+  lazy val maxBlockSize: Int = parametersTable(MaxBlockSizeIncrease)
+
+  /**
+    * Max total computation cost of a block.
+    */
+  lazy val maxBlockCost: Long = parametersTable(MaxBlockCostIncrease)
+
+  lazy val softForkStartingHeight: Option[Height] = parametersTable.get(SoftForkStartingHeight)
+  lazy val softForkVotesCollected: Option[Int] = parametersTable.get(SoftForkVotesCollected)
+
+  lazy val blockVersion: Byte = parametersTable(BlockVersion).toByte
+
+  def update(height: Height, forkVote: Boolean, epochVotes: Seq[(Byte, Int)], votingSettings: VotingSettings): Parameters = {
+    val table1 = updateFork(height, parametersTable, forkVote, epochVotes, votingSettings)
+    val table2 = updateParams(table1, epochVotes, votingSettings.votingLength)
+    Parameters(height, table2)
+  }
+
+  def updateFork(height: Height,
+                 parametersTable: Map[Byte, Int],
+                 forkVote: Boolean,
+                 epochVotes: Seq[(Byte, Int)],
+                 votingSettings: VotingSettings): Map[Byte, Int] = {
+    lazy val votingEpochLength = votingSettings.votingLength
+    lazy val votingEpochs = votingSettings.softForkEpochs
+    lazy val activationEpochs = votingSettings.activationEpochs
+
+    var table = parametersTable
+
+    def activationHeight(startingHeight: Height) =
+      startingHeight + (votingEpochs + activationEpochs) * votingEpochLength
+
+    lazy val votesInPrevEpoch = epochVotes.find(_._1 == SoftFork).map(_._2).getOrElse(0)
+    lazy val votes = votesInPrevEpoch + parametersTable(SoftForkVotesCollected)
+
+    //successful voting - cleaning after activation
+    if(softForkStartingHeight.nonEmpty
+      && height % votingEpochLength == 0
+      && height == softForkStartingHeight.get + votingEpochLength * (votingEpochs + activationEpochs + 1)) {
+
+      val votes = parametersTable(SoftForkVotesCollected)
+
+      if (votes > votingEpochLength * votingEpochs * 9 / 10) {
```

Comment:
`votingEpochLength * votingEpochs * 9 / 10` combination is used quite often, it is better calculate it once inside `votingSettings`. I think the best way is to extract 2 methods to votingSettings:
```
case class VotingSettings(votingLength: Int, softForkEpochs: Int, activationEpochs: Int) {
  def successfulSoftFork(count: Int): Boolean = count > votingLength * softForkEpochs * 9 / 10

  def successfulVote(count: Int): Boolean = count > votingLength / 2
}
```

Reply:
added the methods under a bit different names